src/swapRouter.ts

```diff
@@ -175,17 +179,173 @@ export abstract class SwapRouter {
+    trade: MixedRouteTrade<Currency, Currency, TradeType>,
+    options: SwapOptions,
+    /// @dev we might not need these two flags since that behavior is assumed to happen for mixed route
+    /// they are not used in mixedRouteTrade calldata generation but passed to encodeV3Swap and encodeV2Swap in case of singleHop
+    routerMustCustody: boolean,
+    performAggregatedSlippageCheck: boolean
+  ): string[] {
+    const calldatas: string[] = []
+
+    invariant(trade.tradeType === TradeType.EXACT_INPUT, 'MixedRouteTrades must be exact input')
+
+    for (const { route, inputAmount, outputAmount } of trade.swaps) {
+      const amountIn: string = toHex(trade.maximumAmountIn(options.slippageTolerance, inputAmount).quotient)
+      const amountOut: string = toHex(trade.minimumAmountOut(options.slippageTolerance, outputAmount).quotient)
+
+      // flag for whether the trade is single hop or not
+      const singleHop = route.pools.length === 1
+
+      const recipient =
+        typeof options.recipient === 'undefined' ? MSG_SENDER : validateAndParseAddress(options.recipient)
+      const mixedRouteIsAllV3 = (route: MixedRoute<Currency, Currency>) => {
+        return route.pools.every((pool) => pool instanceof Pool)
+      }
+
+      if (singleHop) {
+        // For single hop, since it isn't really a mixedRoute, we'll just mimic behavior of V3 or V2
+        if (route.pools.every((pool) => pool instanceof Pool)) {
+          /// it's guaranteed that the MixedRouteTrade here is compatible with V3Trade
+          return this.encodeV3Swap(
+            trade as unknown as V3Trade<Currency, Currency, TradeType>,
+            options,
+            routerMustCustody,
+            performAggregatedSlippageCheck
+          )
+        } else {
+          return [
+            ...calldatas,
+            this.encodeV2Swap(
+              trade as unknown as V2Trade<Currency, Currency, TradeType>,
+              options,
+              routerMustCustody,
+              performAggregatedSlippageCheck
+            ),
+          ]
+        }
+      } else {
+        let acc = []
+        let j = 0
+        while (j < route.pools.length) {
+          // seek forward until finding a pool of different type
+          let section = []
+          if (route.pools[j] instanceof Pool) {
+            while (route.pools[j] instanceof Pool) {
+              section.push(route.pools[j])
+              j++
+              if (j === route.pools.length) {
+                // we've reached the end of the route
+                break
+              }
+            }
+            acc.push(section)
+          } else {
+            while (route.pools[j] instanceof Pair) {
+              section.push(route.pools[j])
+              j++
+              if (j === route.pools.length) {
+                // we've reached the end of the route
+                break
+              }
+            }
+            acc.push(section)
+          }
+        }
```

Comment:
Rather than sending every individual pool/pair in the mixed route as a separate swap, we encode consecutive sections of v2 and v3 to send as "mini routes"